File: packaging/agent_evidence_vectors/contextdiscovery.py
"""A validator for a discovery snapshot under draft-arsentev-llm-context-discovery-00.

The draft is an individual Internet-Draft (Informational, September 2026)
defining how a publisher advertises a context file for language-model
consumers and how a consumer resolves it: a well-known URI, a link relation,
and a robots.txt record, with a precedence between them and the robots
exclusion protocol above all three. Its text is vendored beside the corpus that
cites it and pinned by digest; each rule here is bound to one sentence of that
text by a requirement identifier the corpus mints (``LCD-R-nnn``).

A member is a SNAPSHOT: what the origin served on each mechanism, which
resources exist and whether each is an index or a detail resource, what the
robots.txt carries, and what one consumer then did. The publisher-side rows
read the mechanisms and the consumer-side rows read the consumer's record
against them, so one document can answer for either side.

``corpora/contextdiscovery.go`` is the same statement in Go and the parity
test diffs the two.
"""

from __future__ import annotations

from typing import Any

WELL_KNOWN = "/.well-known/llm-context"
REDIRECTS = (301, 302, 307, 308)

R = {n: f"LCD-R-{n:03d}" for n in range(1, 12)}


def _is_str(value: Any) -> bool:
    return isinstance(value, str)


def _is_obj(value: Any) -> bool:
    return isinstance(value, dict)


def _is_int(value: Any) -> bool:
    return isinstance(value, int) and not isinstance(value, bool)


def shape_errors(document: Any) -> list[str]:
    """Every way the document fails to be a discovery snapshot at all."""
    if not _is_obj(document):
        return ["the document is not a JSON object"]
    out: list[str] = []
    if not _is_str(document.get("origin")):
        out.append("the snapshot carries no string origin")
    if not _is_obj(document.get("resources")):
        out.append("the snapshot carries no resources object")
    if not _is_obj(document.get("consumer")):
        out.append("the snapshot carries no consumer object")
    for slot in ("well-known", "link", "robots"):
        if slot in document and document[slot] is not None and not _is_obj(document[slot]):
            out.append(f"{slot} is present and is not an object")
    return out


def _absolute(uri: Any) -> bool:
    return _is_str(uri) and (uri.startswith("https://") or uri.startswith("http://"))


def _origin_of(uri: str) -> str:
    scheme, _, rest = uri.partition("://")
    return scheme + "://" + rest.split("/", 1)[0]


def _path_of(uri: str) -> str:
    _, _, rest = uri.partition("://")
    return "/" + rest.split("/", 1)[1] if "/" in rest else "/"


def _disallowed(uri: str, origin: str, robots: dict[str, Any] | None) -> bool:
    if not _absolute(uri) or _origin_of(uri) != origin or robots is None:
        return False
    rules = [r for r in robots.get("disallow") or [] if _is_str(r) and r]
    return any(_path_of(uri).startswith(rule) for rule in rules)


def _records(robots: dict[str, Any] | None) -> list[Any]:
    if robots is None:
        return []
    out = []
    for record in robots.get("records") or []:
        if not _is_obj(record) or not _is_str(record.get("name")):
            continue
        if record["name"].lower() == "llm-context":
            out.append(record.get("value"))
    return out


def _well_known_target(document: dict[str, Any], out: set[str]) -> str | None:
    """The URI the well-known mechanism advertises, or None when it advertises nothing."""
    response = document.get("well-known")
    if response is None:
        return None
    status = response.get("status")
    if status == 404:
        return None
    if _is_int(status) and 200 <= status < 300:
        content_type = response.get("content-type")
        if _is_str(content_type) and content_type.lower().startswith("text/html"):
            out.add(R[1])
        if response.get("negotiated") is True and not _is_str(response.get("content-language")):
            out.add(R[10])
        return str(document["origin"]) + WELL_KNOWN
    if status in REDIRECTS and _absolute(response.get("location")):
        return str(response["location"])
    out.add(R[3])
    return None


def _rows_publisher(document: dict[str, Any], out: set[str]) -> str | None:
    """The publisher rows, returning the URI the precedence rule selects."""
    origin = document["origin"]
    resources = document["resources"]
    robots = document.get("robots")
    link: Any = document.get("link")
    link_target = link.get("target") if _is_obj(link) else None
    well_known = _well_known_target(document, out)
    robots_targets: list[str] = []
    for value in _records(robots):
        if not _absolute(value):
            out.add(R[4])
            continue
        if _disallowed(value, origin, robots):
            out.add(R[5])
        robots_targets.append(value)
    targets = [t for t in [link_target, well_known, *robots_targets] if _is_str(t)]
    for target in targets:
        resource = resources.get(target)
        if _is_obj(resource) and resource.get("role") == "detail":
            out.add(R[2])
            break
    if _is_str(link_target):
        return link_target
    if well_known is not None:
        return well_known
    return robots_targets[0] if robots_targets else None


def _rows_consumer(document: dict[str, Any], selected: str | None, out: set[str]) -> None:
    origin = document["origin"]
    resources = document["resources"]
    robots = document.get("robots")
    consumer = document["consumer"]
    resolved = consumer.get("resolved")
    if resolved is not None and resolved != selected:
        out.add(R[6])
    retrieved = [u for u in consumer.get("retrieved") or [] if _is_str(u)]
    known = [u for u in retrieved if _is_obj(resources.get(u))]
    indexes = [u for u in known if resources[u].get("role") == "index"]
    if len(indexes) > 1:
        out.add(R[7])
    foreign = _absolute(resolved) and _origin_of(str(resolved)) != origin
    if foreign and consumer.get("attributed-to") == origin:
        out.add(R[8])
    ceiling = consumer.get("ceiling-octets")
    if not _is_int(ceiling) or ceiling <= 0:
        out.add(R[9])
    if any(_disallowed(u, origin, robots) for u in retrieved):
        out.add(R[11])


def rejections(document: dict[str, Any]) -> list[str]:
    """The requirement identifiers this snapshot is rejected under, sorted."""
    out: set[str] = set()
    selected = _rows_publisher(document, out)
    _rows_consumer(document, selected, out)
    return sorted(out)

File: corpora/contextdiscovery.go
package corpora

import (
	"sort"
	"strings"
)

// lcdRejections judges a discovery snapshot under
// draft-arsentev-llm-context-discovery-00. It is the Go statement of
// packaging/agent_evidence_vectors/contextdiscovery.py: the same sentences,
// bound to the same LCD-R-nnn identifiers, so the parity test can diff the two.

const lcdWellKnown = "/.well-known/llm-context"

var lcdRedirects = map[int64]bool{301: true, 302: true, 307: true, 308: true}

func lcdR(n int) string { return "LCD-R-" + pad3(n) }

type lcdRejects map[string]bool

func (r lcdRejects) add(n int) { r[lcdR(n)] = true }

func lcdShapeErrors(value any) []string {
	document, ok := value.(map[string]any)
	if !ok {
		return []string{"the document is not a JSON object"}
	}
	var out []string
	if !isStr(document["origin"]) {
		out = append(out, "the snapshot carries no string origin")
	}
	if !isObj(document["resources"]) {
		out = append(out, "the snapshot carries no resources object")
	}
	if !isObj(document["consumer"]) {
		out = append(out, "the snapshot carries no consumer object")
	}
	for _, slot := range []string{"well-known", "link", "robots"} {
		if raw, present := document[slot]; present && raw != nil && !isObj(raw) {
			out = append(out, slot+" is present and is not an object")
		}
	}
	return out
}

func lcdAbsolute(v any) bool {
	s, ok := v.(string)
	return ok && (strings.HasPrefix(s, "https://") || strings.HasPrefix(s, "http://"))
}

func lcdOrigin(uri string) string {
	scheme, rest, _ := strings.Cut(uri, "://")
	host, _, _ := strings.Cut(rest, "/")
	return scheme + "://" + host
}

func lcdPath(uri string) string {
	_, rest, _ := strings.Cut(uri, "://")
	_, path, found := strings.Cut(rest, "/")
	if !found {
		return "/"
	}
	return "/" + path
}

func lcdDisallowed(uri string, origin string, robots map[string]any) bool {
	if !lcdAbsolute(uri) || lcdOrigin(uri) != origin || robots == nil {
		return false
	}
	rules, _ := robots["disallow"].([]any)
	for _, raw := range rules {
		rule, ok := raw.(string)
		if ok && rule != "" && strings.HasPrefix(lcdPath(uri), rule) {
			return true
		}
	}
	return false
}

func lcdRecords(robots map[string]any) []any {
	if robots == nil {
		return nil
	}
	records, _ := robots["records"].([]any)
	var out []any
	for _, raw := range records {
		record, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		name, ok := record["name"].(string)
		if !ok {
			continue
		}
		if strings.ToLower(name) == "llm-context" {
			out = append(out, record["value"])
		}
	}
	return out
}

// lcdWellKnownTarget is the URI the well-known mechanism advertises, and
// whether it advertises one at all.
func lcdWellKnownTarget(document map[string]any, out lcdRejects) (string, bool) {
	response, ok := document["well-known"].(map[string]any)
	if !ok {
		return "", false
	}
	status, isInt := intValue(response["status"])
	if isInt && status == 404 {
		return "", false
	}
	if isInt && status >= 200 && status < 300 {
		if contentType, ok := response["content-type"].(string); ok && strings.HasPrefix(strings.ToLower(contentType), "text/html") {
			out.add(1)
		}
		if negotiated, _ := response["negotiated"].(bool); negotiated && !isStr(response["content-language"]) {
			out.add(10)
		}
		origin, _ := document["origin"].(string)
		return origin + lcdWellKnown, true
	}
	if isInt && lcdRedirects[status] && lcdAbsolute(response["location"]) {
		return response["location"].(string), true
	}
	out.add(3)
	return "", false
}

func lcdRowsPublisher(document map[string]any, out lcdRejects) (string, bool) {
	origin, _ := document["origin"].(string)
	resources, _ := document["resources"].(map[string]any)
	robots, _ := document["robots"].(map[string]any)
	link, _ := document["link"].(map[string]any)
	linkTarget, hasLink := link["target"].(string)
	wellKnown, hasWellKnown := lcdWellKnownTarget(document, out)
	var robotsTargets []string
	for _, value := range lcdRecords(robots) {
		if !lcdAbsolute(value) {
			out.add(4)
			continue
		}
		target := value.(string)
		if lcdDisallowed(target, origin, robots) {
			out.add(5)
		}
		robotsTargets = append(robotsTargets, target)
	}
	var targets []string
	if hasLink {
		targets = append(targets, linkTarget)
	}
	if hasWellKnown {
		targets = append(targets, wellKnown)
	}
	targets = append(targets, robotsTargets...)
	for _, target := range targets {
		if resource, ok := resources[target].(map[string]any); ok {
			if role, _ := resource["role"].(string); role == "detail" {
				out.add(2)
				break
			}
		}
	}
	switch {
	case hasLink:
		return linkTarget, true
	case hasWellKnown:
		return wellKnown, true
	case len(robotsTargets) > 0:
		return robotsTargets[0], true
	}
	return "", false
}

func lcdRowsConsumer(document map[string]any, selected string, selectedOK bool, out lcdRejects) {
	origin, _ := document["origin"].(string)
	resources, _ := document["resources"].(map[string]any)
	robots, _ := document["robots"].(map[string]any)
	consumer, _ := document["consumer"].(map[string]any)
	if resolved, present := consumer["resolved"]; present && resolved != nil {
		s, isString := resolved.(string)
		if !isString || !selectedOK || s != selected {
			out.add(6)
		}
	}
	rawRetrieved, _ := consumer["retrieved"].([]any)
	var retrieved []string
	for _, raw := range rawRetrieved {
		if s, ok := raw.(string); ok {
			retrieved = append(retrieved, s)
		}
	}
	indexes := 0
	for _, uri := range retrieved {
		if resource, ok := resources[uri].(map[string]any); ok {
			if role, _ := resource["role"].(string); role == "index" {
				indexes++
			}
		}
	}
	if indexes > 1 {
		out.add(7)
	}
	resolved, _ := consumer["resolved"].(string)
	foreign := lcdAbsolute(consumer["resolved"]) && lcdOrigin(resolved) != origin
	if foreign && jsonEqual(consumer["attributed-to"], origin) {
		out.add(8)
	}
	if ceiling, ok := intValue(consumer["ceiling-octets"]); !ok || ceiling <= 0 {
		out.add(9)
	}
	for _, uri := range retrieved {
		if lcdDisallowed(uri, origin, robots) {
			out.add(11)
			break
		}
	}
}

func lcdRejections(document map[string]any) []string {
	out := lcdRejects{}
	selected, ok := lcdRowsPublisher(document, out)
	lcdRowsConsumer(document, selected, ok, out)
	ids := make([]string, 0, len(out))
	for id := range out {
		ids = append(ids, id)
	}
	sort.Strings(ids)
	return ids
}
